### PR TITLE
Implement cpad w/ keyboard input into runtime

### PR DESCRIPTION
**Pull request by ManDude**

### CODE REVIEW

**Comment by ManDude** [comment]
This is a very basic implementation of the input system for the pc port. It maps buttons on the keyboard to cpad buttons used by the game. It **requires** running with a display on, and inputs are recognized in the display window. By default, the following keymapping is used:
- X to Z key
- Square to X key
- Triangle to S key
- Circle to A key
- D-pad to the arrow keys
The rest are unmapped.

There's some functions for debugging and testing this implementation. If you want to run the pc pad util functions, you need to `(ml "goal_src/pc_debug/pc-pad-utils.gc")` first.
You can run a placeholder button mapping util ***for cpad 0*** on the runtime by running `(pc-pad-input-start)` and following the onscreen instructions. You simply press the keys you want to attach to each button. When the mapping is complete, it is saved to memory and to the config. This is implemented with an actual GOAL process! Pretty neat.
If you wish to cancel out of the button mapping util, just press Escape or run `(pc-pad-input-stop)`. This also means you can't map any buttons to Escape.
You can check the current button mappings by running `(pc-pad-show-start)`. This prints the button mapping to `*stdout*` on your left (it's too much text however and cpad 1 is cut off). `(pc-pad-show-stop)` will kill that process.

Graphics (& button mapping) settings are saved to a `game_config` folder in the project root whenever you finish the button mapping util.

File: game/graphics/pipelines/opengl.cpp [hunk 1]
```diff
@@ -77,11 +80,16 @@ static int gl_init() {
...
+  // request an OpenGL 3.3 Core context
   glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3);  // 3.3
   glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 3);
-  // glfwWindowHint(GLFW_OPENGL_DEBUG_CONTEXT, GLFW_TRUE);           // debug
   glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);  // core profile, not compat
+  // debug check
+  if (settings.debug) {
+    glfwWindowHint(GLFW_OPENGL_DEBUG_CONTEXT, GLFW_TRUE);
+  } else {
+    glfwWindowHint(GLFW_OPENGL_DEBUG_CONTEXT, GLFW_FALSE);
```
Comment (water111) [inline game/graphics/pipelines/opengl.cpp:91]: good idea to make this an option

File: goal_src/engine/ps2/pad.gc [hunk 2]
```diff
@@ -30,24 +30,41 @@
...
     (square   15)
     )
-    
+
+(defenum pad-type
+  (normal 4)
+  (analog 5)
+  (dualshock 7)
+  (negcon 2)
+  (namco-gun 6)
```
Comment (water111) [inline goal_src/engine/ps2/pad.gc:39]: we should add support for the namco gun :)

File: goal_src/pc_debug/pc-pad-utils.gc [hunk 3]
```diff
@@ -0,0 +1,245 @@
+;;-*-Lisp-*-
+(in-package goal)
```
Comment (water111) [inline goal_src/pc_debug/pc-pad-utils.gc:2]: this file is awesome, it's great to see that we can write our own GOAL code with states and rendering and have it actually work!

File: goal_src/pc_debug/pc-pad-utils.gc [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+  )
+
+(defun-debug pc-pad-show-start ()
+  "Start the PC port pad debug display"
+
+  (if (not (handle->process (-> *pc-pad-proc-list* show)))
+    (let ((procp
+      (make-function-process *nk-dead-pool* *active-pool* pc-pad-proc :name 'pc-pad-show
+        (lambda :behavior pc-pad-proc ()
+          (stack-size-set! (-> self main-thread) 512)
```
Comment (water111) [inline goal_src/pc_debug/pc-pad-utils.gc:76]: I'm kind of surprised we needed more than the default stack size here.  I'm guessing you hit the limit when this was smaller?

**Review by water111 — APPROVED** [review]
awesome!

**Pull request merged**